File: apps/go-api/internal/presence/presence.go
package presence

import (
	"context"
	"strconv"
	"time"

	goredis "github.com/redis/go-redis/v9"
)

const (
	presenceKey    = "users:online"
	presenceTTLSec = 60
	ttlMs          = presenceTTLSec * 1000
)

// Track marks a user as online for 60 seconds. Fire-and-forget.
// Matches TS presence.ts trackPresence() exactly:
// MULTI { ZADD users:online <now> <userId>, ZREMRANGEBYSCORE users:online -inf <cutoff> }
func Track(ctx context.Context, rdb *goredis.Client, userID string) error {
	now := time.Now().UnixMilli()
	cutoff := now - int64(ttlMs)

	pipe := rdb.TxPipeline()
	pipe.ZAdd(ctx, presenceKey, goredis.Z{Score: float64(now), Member: userID})
	pipe.ZRemRangeByScore(ctx, presenceKey, "-inf", strconv.FormatInt(cutoff, 10))
	_, err := pipe.Exec(ctx)
	return err
}

// CountOnline returns the number of users active in the last 60 seconds.
// Matches TS presence.ts countOnlineUsers() exactly:
// ZREMRANGEBYSCORE then ZCARD.
func CountOnline(ctx context.Context, rdb *goredis.Client) (int64, error) {
	cutoff := time.Now().UnixMilli() - int64(ttlMs)
	if err := rdb.ZRemRangeByScore(ctx, presenceKey, "-inf", strconv.FormatInt(cutoff, 10)).Err(); err != nil {
		return 0, err
	}
	return rdb.ZCard(ctx, presenceKey).Result()
}

File: apps/api/src/lib/presence.ts
import type Redis from 'ioredis';

const PRESENCE_TTL_SEC = 60;
const PRESENCE_SORTED_SET_KEY = 'users:online';
const TTL_MS = PRESENCE_TTL_SEC * 1000;

/** Fire-and-forget: mark a user as active for the next 60 seconds. */
export function trackPresence(userId: string, redis: Redis): Promise<unknown> {
  const now = Date.now();
  const cutoff = now - TTL_MS;
  return redis
    .multi()
    .zadd(PRESENCE_SORTED_SET_KEY, String(now), userId)
    .zremrangebyscore(PRESENCE_SORTED_SET_KEY, '-inf', String(cutoff))
    .exec();
}

/** Count active users in O(log N) using a heartbeat sorted set. */
export async function countOnlineUsers(redis: Redis): Promise<number> {
  const cutoff = Date.now() - TTL_MS;
  await redis.zremrangebyscore(PRESENCE_SORTED_SET_KEY, '-inf', String(cutoff));
  return redis.zcard(PRESENCE_SORTED_SET_KEY);
}
